File: gtci/05-cyclic-sort/python3/5-find-the-corrupt-pair.py
class Solution:
    def findNumbers(self, nums):
        i = 0
        while i < len(nums):
            correct_position = nums[i] - 1
            if nums[i] != nums[correct_position]:
                nums[i], nums[correct_position] = nums[correct_position], nums[i]
            else:
                i += 1

        for i in range(len(nums)):
            if i + 1 != nums[i]:
                return [nums[i], i + 1]

        return [-1, -1]


if __name__ == "__main__":
    sol = Solution()

    assert sol.findNumbers([3, 1, 2, 5, 2]) == [2, 4]
    assert sol.findNumbers([3, 1, 2, 3, 6, 4]) == [3, 5]

    print("All test cases passed.")

File: gtci/05-cyclic-sort/cpp/5-find-the-corrupt-pair.cpp
#include <cassert>
#include <iostream>
#include <string>
#include <utility>
#include <vector>

using namespace std;

class Solution {
   public:
    vector<int> findNumbers(vector<int>&& nums) {
        size_t i = 0;
        while (i < nums.size()) {
            size_t correctPosition = nums[i] - 1;

            if (nums[i] != nums[correctPosition]) {
                std::swap(nums[i], nums[correctPosition]);
            } else {
                i++;
            }
        }

        for (i = 0; i < nums.size(); i++) {
            if (i + 1 != nums[i])
                return vector<int>{nums[i], static_cast<int>(i + 1)};
        }
        return vector<int>{-1, -1};
    }
};

int main() {
    Solution sol;

    assert(sol.findNumbers({3, 1, 2, 5, 2}) == vector<int>({2, 4}));
    assert(sol.findNumbers({3, 1, 2, 3, 6, 4}) == vector<int>({3, 5}));

    cout << "All test cases passed." << endl;

    return 0;
}
